lasagne/layers/normalization.py

```diff
@@ -101,3 +115,233 @@ def get_output_for(self, input, **kwargs):
+    mean : Theano shared variable, expression, numpy array, or callable
+        Initial value, expression or initializer for :math:`\\mu`. Must match
+        the incoming shape, skipping all axes in `axes`.
+        See :func:`lasagne.utils.create_param` for more information.
+    var : Theano shared variable, expression, numpy array, or callable
+        Initial value, expression or initializer for :math:`\\sigma^2`. Must
+        match the incoming shape, skipping all axes in `axes`.
+        See :func:`lasagne.utils.create_param` for more information.
+    **kwargs
+        Any additional keyword arguments are passed to the :class:`Layer`
+        superclass.
+
+    Notes
+    -----
+    This layer should be inserted between a linear transformation (such as a
+    :class:`DenseLayer`, or :class:`Conv2DLayer`) and its nonlinearity. The
+    convenience function :func:`batch_norm` modifies an existing layer to
+    insert batch normalization in front of its nonlinearity.
+
+    The behavior can be controlled by passing keyword arguments to
+    :func:`lasagne.layers.get_output()` when building the output expression
+    of any network containing this layer.
+
+    During training, [1]_ normalize each input mini-batch by its statistics
+    and update an exponential moving average of the statistics to be used for
+    validation. This can be achieved by passing ``deterministic=False``.
+    For validation, [1]_ normalize each input mini-batch by the stored
+    statistics. This can be achieved by passing ``deterministic=True``.
+
+    For more fine-grained control, ``batch_norm_update_averages`` can be passed
+    to update the exponential moving averages (``True``) or not (``False``),
+    and ``batch_norm_use_averages`` can be passed to use the exponential moving
+    averages for normalization (``True``) or normalize each mini-batch by its
+    own statistics (``False``). These settings override ``deterministic``.
+
+    Note that for testing a model after training, [1]_ replace the stored
+    exponential moving average statistics by fixing all network weights and
+    re-computing average statistics over the training data in a layerwise
+    fashion. This is not part of the layer implementation.
+
+    See also
+    --------
+    batch_norm : Convenience function to apply batch normalization to a layer
+
+    References
+    ----------
+    .. [1]: Ioffe, Sergey and Szegedy, Christian (2015):
+            Batch Normalization: Accelerating Deep Network Training by Reducing
+            Internal Covariate Shift. http://arxiv.org/abs/1502.03167.
+    """
+    def __init__(self, incoming, axes='auto', epsilon=1e-4, alpha=0.5,
+                 nonlinearity=None, mode='low_mem',
+                 beta=init.Constant(0), gamma=init.Constant(1),
+                 mean=init.Constant(0), var=init.Constant(1), **kwargs):
+        super(BatchNormLayer, self).__init__(incoming, **kwargs)
```

Comment:
I don't think so. I'm not actually sure whether it should have a `nonlinearity` at all.

In hindsight, I think the best solution would have been not to have a `nonlinearity` keyword in any of the layers, and make the nonlinearities Layer classes as well. That would make it much more straightforward to support parameterizable nonlinearities, and to insert batch normalization between a linear transformation and its nonlinearity. It's too late for that, though.

I added a `nonlinearity` keyword here because usually, you will need a nonlinearity right after the batch normalization, but maybe it should be removed completely? I'd find it confusing to have the `BatchNormLayer` apply a linear rectifier by default, as that's not really the purpose of batch normalization.